src/vec/mod.rs

```diff
@@ -0,0 +1,1084 @@
+///
+/// Compare [`std::Vec`].
+pub struct Vec<T> {
+  ptr: NonNull<T>,
+  capacity: usize,
+  len: usize,
+
+  // Scratch space for a drop flag to use inside of an iterator.
+  // This does not affect synchronization, since it is only used when we
+  // have a &mut Vec anyways.
+  //
+  // This is *not* used for trapping the overall vector! A Vec<T> owns its
+  // storage and can be forgotten safely.
+  drop_flag: QuietFlag,
+}
+
+// SAFETY: std::Vec<T> is Send + Sync; see comment on scratch_space above.
+unsafe impl<T: Send> Send for Vec<T> {}
+unsafe impl<T: Sync> Sync for Vec<T> {}
+
+impl<T> Vec<T> {
+  /// True if the element type is a ZST.
+  ///
+  /// This constant is used to short-circuit certain kinds of logic that don't
+  /// really make sense with ZSTs, which is roughly how `Vec` works too. The
+  /// following things have been changed:
+  /// - Allocation always succeeds.
+  /// - Capacity is infinite.
+  /// - Insertion (and pushing) just increment the length.
+  /// - The pointer is always dangling.
+  /// - Removal always returns the same pointer value; we formulate this as
+  ///   there being infinitely-many disjoint objects at the same address.
+  const ZST: bool = mem::size_of::<T>() == 0;
+
+  const MIN_CAPACITY: usize = 16;
+
+  /// Constructs a new, empty `Vec<T>`.
+  ///
+  /// Compare [`std::Vec::new()`].
+  pub fn new() -> Self {
+    Self {
+      ptr: NonNull::dangling(),
+      capacity: if Self::ZST { usize::MAX } else { 0 },
+      len: 0,
+      drop_flag: QuietFlag::new(),
+    }
+  }
+
+  /// Constructs a new, empty `Vec<T>` with the specified capacity.
+  ///
+  /// Compare [`std::Vec::with_capacity()`].
+  ///
+  /// # Panics
+  ///
+  /// Panics if the requested allocation size (in bytes) overflows an isize.
+  pub fn with_capacity(capacity: usize) -> Self {
+    if capacity == 0 || Self::ZST {
+      return Self::new();
+    }
+
+    let layout =
+      Layout::array::<T>(capacity).expect("overflow in layout calculation");
+    let ptr = unsafe { alloc::alloc::alloc(layout) };
+    assert!(!ptr.is_null(), "allocation failure");
+
+    Self {
+      ptr: unsafe { NonNull::new_unchecked(ptr as *mut T) },
+      capacity,
+      len: 0,
+      drop_flag: QuietFlag::new(),
+    }
+  }
+
+  /// Creates a `Vec<T>` using the raw components of another vector.
+  ///
+  /// Compare [`std::Vec::from_raw_parts()`].
+  ///
+  /// # Safety
+  ///
+  /// This function will blindly re-assemble a `Vec<T>` without thinking. The
+  /// usual caveats apply.
+  pub unsafe fn from_raw_parts(
+    ptr: *mut T,
+    length: usize,
+    capacity: usize,
+  ) -> Self {
+    Self {
+      ptr: NonNull::new_unchecked(ptr),
+      len: length,
+      capacity,
+      drop_flag: QuietFlag::new(),
+    }
+  }
+
+  /// Creates a new `Vec<T>` using the constructors returned by `iter`.
+  ///
+  /// Due to trait coherence issues, this cannot be exposed via `.collect()`.
```

Comment:
What are the coherence issues here? I've copied the bounds and the `FromIterator` impl seems to be working: [[play]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0964aaf32471d4b7a0fc5fc2abdfd8f8), [[play_fancy]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f4d3cc363a39aca56c2987c335ca0945).